src/au3wrap/internal/nonblockingprogressdialog.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * Audacity: A Digital Audio Editor
+ */
+#include "nonblockingprogressdialog.h"
+
+#include "iinteractive.h"
+#include "modularity/ioc.h"
+#include "async/asyncable.h"
+
+#include "libraries/lib-exceptions/AudacityException.h"
+
+#include <QEventLoop>
+
+#include <thread>
+
+namespace au::effects {
+namespace {
+using namespace NonBlockingProgressDialog;
+
+class DialogImpl : public muse::async::Asyncable
+{
+    muse::Inject<muse::IInteractive> interactive;
+
+public:
+    muse::Ret execute(Task task, std::string title, muse::Ret::Code errorCode)
+    {
+        QEventLoop loop;
+        muse::Ret ret;
+
+        m_museProgress.finished().onReceive(this, [&loop, &ret](const muse::RetVal<muse::Val>& retval) {
+            ret = retval.ret;
+            loop.quit();
+        });
+
+        // Make sure that the progress is started before the thread starts being executed.
+        interactive()->showProgress(title, &m_museProgress);
+
+        std::thread thread([&] {
+            DialogImpl::Au3Progress au3Progress { m_museProgress };
+
+            try {
+                if (!task(au3Progress)) {
+                    m_museProgress.finished().send(muse::make_ret(errorCode));
+                } else if (!m_museProgress.isCanceled()) {
+                    m_museProgress.finished().send(muse::make_ok());
+                }
+            } catch (::AudacityException& e) {
+                std::string message;
+                if (const auto box = dynamic_cast<::MessageBoxException*>(&e)) {
+                    message = box->ErrorMessage().Translation().ToStdString();
+                }
+                m_museProgress.finished().send(muse::make_ret(errorCode, message));
```

Comment:
If there will be non audacity exception we will crash / loop never exit.

Reply:
Yeah, we do not catch all exceptions (we probably should) on the main thread either. The difference is that even if you wrap the call to `NonBlockingProgressDialog::execute` in a catch-all, the inner thread will still crash.

Since this class is a general purpose progress task runner, it should not crash on random exceptions